src/hooks/useQueryState.js

```diff
@@ -0,0 +1,163 @@
+  };
+  const defaultValues = {
+    SingleValue: '',
+    Array: [],
+    Boolean: false,
+  };
+  const type = types[typeString];
+  if (type === undefined)
+    throw new Error(`type must be one of the given types: ${Object.values(types)}`);
+  const defaultVal = defaultValues[type];
+  const [state, setState] = useState(initial ?? defaultVal);
+  const history = useHistory();
+
+  /**
+   * Anytime the state changes, setURLParam is invoked
+   * This function updates the URL with the state (if changed)
+   */
+  useEffect(() => {
+    const setURLParam = async () => {
+      let urlParams = new URLSearchParams(history.location.search);
+
+      // if the url value is already the same as the state value -> don't set the url again
+      // note: must include toString in defaultVal.toString() in order to check empty arrays
+      if (
+        (urlParams.get(key) ?? defaultVal.toString()) === state.toString() ||
+        (state === true && urlParams.get(key) === '')
+      )
+        return;
+
+      // key with no value OR boolean key set to false -> absent from url, push empty
+      if (!state || state === '' || state.length === 0 || state === false) urlParams.delete(key);
+      // boolean or value key is present in url with its current value
+      else urlParams.set(key, state.toString());
+
+      // boolean keys set to true don't need '=value'
+      urlParams = urlParams
+        .toString()
+        .replace(/=true|=$/g, '')
+        .replace(/=&/, '&');
+
+      history.push('?' + urlParams);
+    };
+    setURLParam();
+    /*********************** ESLint Disable Justification: *******************************/
+    // ESLint wants to include `defaultVal, key, history` as dependencies but to prevent
+    // extra loops of ~ url setting state setting the url ~ we do not include.
+    // The only time we want to set the URL manually is when the state changes
+    /*************************************************************************************/
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [state]);
+
+  /**
+   * Anytime the url is changed, loadURLParams is invoked
+   * This function pulls the value of the key from the url
+   *   and updates the state with it (if changed)
+   *
+   * This is a useLayoutEffect rather than useEffect in order to prioritize loadURLParams;
+   *   this is necessary for preventing the race condition where (for ex.) the user clicks 'back'
+   *   twice quickly and the first 'back' sets the URL params before the second 'back' can
+   *   load the accurate new URL param. This is still not totally perfect but only breaks when
+   *   a user really messes with the navigation significantly and quickly.
+   * Comparison of the two effects: https://kentcdodds.com/blog/useeffect-vs-uselayouteffect
+   */
+  useLayoutEffect(() => {
+    const loadURLParams = async () => {
+      const urlParams = new URLSearchParams(history.location.search);
+      let urlValue = urlParams.get(key);
```

Comment:
Does this load array values properly? I only ask because I know that there's also `urlParams.getAll(key)`.

Reply:
Ah, I see. Both ways are actually equally valid methods for storing an array in the query string, since there is no agreed upon standard. In fact, if it would be simpler to do your latter example, we should probably do that.